# Contents<a id='back'></a>
* [Introduction](#intro)
* [Data Review](#data_review)
    * [Getting General Information About the Data](#open_file)
    * [Data Exploration](#data_explore)
    * [Conclusion](#conclusion1)
* [Data Preprocessing](#processing_data)
    * [Data Transformation](#data_transform)
    * [Fixing Duplicate Data](#duplicate)
    * [Filling Missing Data](#missing_values)
    * [Data Categorization](#data_category)
* [Hypothesis Testing](#hypothesis)
* [Overall Conclusion](#final)

# Introduction: Analyzing Borrower Default Risk <a id='intro'></a>

The project's goal is to prepare a report for the credit division of a bank. We will investigate the impact of a borrower's marital status and the number of children they have on the probability of loan default. The bank already has some data on customer creditworthiness.

This report will be considered when making **credit assessments** for potential borrowers. **Credit assessment** is used to evaluate the ability of prospective borrowers to repay their loans.

This analysis aims to determine whether there is a relationship between having children, marital status, income level, and the purpose of the loan on the probability of default.

### Objectives

We will test four hypotheses:

1. The relationship between having children and an individual's loan default rate.
2. Whether there is a relationship between marital status and the probability of loan default.
3. Whether an individual's income level can influence loan default.
4. Whether the purpose of the loan affects the likelihood of loan default.

### Steps

The customer data is stored in two different file paths: `'/datasets/credit_scoring_eng.csv'` for the internal file path, and [`'credit_scoring_eng.csv'`](https://raw.githubusercontent.com/milawidyalestari/data-analyst-ml-project/project2-borrower-default-risk/credit_scoring_eng.csv) for the external file path. The quality of the data is not yet known, so we will first review the dataset. Then, we will perform data preprocessing. After completing these steps, we will test our hypotheses.

This project will be divided into three stages:
1. Data Review
2. Data Preprocessing
3. Hypothesis Testing

[Back to contents](#back)

## Stage 1. Data Review <a id= 'data_review'></a>

This stage aims to review the data quality and gather all information related to individuals applying for credit.

### Open the data file and read its general information <a id= 'open_file'></a>

First, we will import the necessary libraries for data processing. We will use the pandas library for data review and cleaning.

Loading all libraries and adding the data:

In [1]:
# Load all libraries
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Load the data
path = 'https://raw.githubusercontent.com/milawidyalestari/data-analyst-ml-project/project2-borrower-default-risk/credit_scoring_eng.csv'
try:
    creditScore = pd.read_csv(path)
except:
    creditScore = pd.read_csv('/datasets/credit_scoring_eng.csv')

### Data Exploration <a id='data_explore'></a>

**Data Description**
- `children` - number of children in the family
- `days_employed` - client's work experience in days
- `dob_years` - client's age in years
- `education` - client's education level
- `education_id` - identifier for the client's education level
- `family_status` - marital status
- `family_status_id` - identifier for the client's marital status
- `gender` - client's gender
- `income_type` - type of employment
- `debt` - whether the client has ever defaulted on a loan
- `total_income` - monthly income
- `purpose` - purpose of obtaining the loan

Reviewing data size:

In [3]:
# Let's see how many rows and columns our dataset has
creditScore.shape

(21525, 12)

Display the first 15 rows:

In [4]:
# Let's display the first N rows.

creditScore.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   

        family_status  family_status_id gender income_type  debt  \
0             married                 0      F    employee     0   
1             married                 0      F    employee     0   
2             married                 0      M    employee     0   
3             married                 0      M    employee     0   
4   civil partnership                 1      F     retiree     0   
5   civil partnership                 1      M    business     0   
6             married                 0      F    business     0   
7             married                 0      M    employee     0   
8   civil partnership                 1      F    employee     0   
9             married                 0      M    employee     0   
10            married                 0      M    business     0   
11            married                 0      F    employee     0   
12  civil partnership                 1      M     retiree     0   
13            married                 0      F    employee     0   
14  civil partnership                 1      F    business     1   

    total_income                              purpose  
0      40620.102                purchase of the house  
1      17932.802                         car purchase  
2      23341.752                purchase of the house  
3      42820.568              supplementary education  
4      25378.572                    to have a wedding  
5      40922.170                purchase of the house  
6      38484.156                 housing transactions  
7      21731.829                            education  
8      15337.093                     having a wedding  
9      23108.150  purchase of the house for my family  
10     18230.959                      buy real estate  
11     12331.077           buy commercial real estate  
12           NaN                    to have a wedding  
13     20873.317                         car purchase  
14     26420.466          buy residential real estate

There is 1 column with a negative value, and there is a null value at index 12. We will conduct further investigation.

Getting dataset information:

In [5]:
# Get data information
creditScore.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


You can see that there are 2 columns that actually have null values. Next, we will find out if the missing values in both columns have symmetric missing values or not.

In [6]:
# Let's view the table filtered by the first column that contains missing data.
null_rows = creditScore.loc[creditScore['days_employed'].isnull()]
null_rows

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [7]:
# Let's view the total number of rows in the filtered dataset.
null_rows.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Alright, the number of null rows in the filtered table turns out to be 2174 rows. We will compare this with the number of null values in our original table with the condition `.isnull()`

In [8]:
# Let's apply some conditions to filter the data and view the number of rows in the original table.
df_null = creditScore[(creditScore['days_employed'].isnull()) & (creditScore['total_income'].isnull())]
df_null

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [9]:
# To know the percentage of missing values in the entire original dataset.
data_null_percent = str(round((len(df_null) / len(creditScore)) * 100, 1)) + '%'
data_null_percent

'10.1%'

### Preliminary Conclusion

The comparison between the filtered dataset and the original dataset with the condition `.isnull()` reveals a similar data distribution. This indicates that all missing null values may occur randomly. Also, note that the null values are symmetrically missing. And approximately 10% of the data actually consists of null values. That's a significant amount of missing data for us. So we need to fill in these values using alternative values.

Next, to further ensure whether the data is missing at random or due to some specific pattern, we will find out the total number of null values without filtering on the table.

In [10]:
# Let's examine customers who lack data on identified characteristics and columns with missing values.
df_null.head(10)

children  days_employed  dob_years            education  education_id  \
12         0            NaN         65  secondary education             1   
26         0            NaN         41  secondary education             1   
29         0            NaN         63  secondary education             1   
41         0            NaN         50  secondary education             1   
55         0            NaN         54  secondary education             1   
65         0            NaN         21  secondary education             1   
67         0            NaN         52    bachelor's degree             0   
72         1            NaN         32    bachelor's degree             0   
82         2            NaN         50    bachelor's degree             0   
83         0            NaN         52  secondary education             1   

        family_status  family_status_id gender    income_type  debt  \
12  civil partnership                 1      M        retiree     0   
26            married                 0      M  civil servant     0   
29          unmarried                 4      F        retiree     0   
41            married                 0      F  civil servant     0   
55  civil partnership                 1      F        retiree     1   
65          unmarried                 4      M       business     0   
67            married                 0      F        retiree     0   
72            married                 0      M  civil servant     0   
82            married                 0      F       employee     0   
83            married                 0      M       employee     0   

    total_income                                   purpose  
12           NaN                         to have a wedding  
26           NaN                                 education  
29           NaN                    building a real estate  
41           NaN                  second-hand car purchase  
55           NaN                         to have a wedding  
65           NaN  transactions with commercial real estate  
67           NaN       purchase of the house for my family  
72           NaN  transactions with commercial real estate  
82           NaN                                   housing  
83           NaN                                   housing

In [11]:
# Check the distribution
data_count = df_null["income_type"].value_counts()
data_percent = df_null["income_type"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : data_percent, "count" : data_count}).rename_axis("unique")

percent%  count
unique                       
employee        50.83%   1105
business        23.37%    508
retiree          19.0%    413
civil servant    6.76%    147
entrepreneur     0.05%      1

### Possible Causes of Data Loss

This is the distribution of the dataframe that has null values **(df_null)** categorized based on the column `'income_type'`. The missing values in the columns `'days_employed'` and `'total_income'` are known to be missing in the following categories:

- employee: 1105 rows
- business: 508 rows
- retiree: 413 rows
- civil servant: 147 rows
- entrepreneur: 1 row

In [12]:
# Checking the distribution across the entire dataset
df_count = df_null['family_status'].value_counts()
df_percent = df_null['family_status'].value_counts(normalize=True).mul(100).round(2).astype(str)+'%'
pd.DataFrame(data={'percent(%)':df_percent, 'count':df_count}).rename_axis('unique')

percent(%)  count
unique                             
married                56.9%   1237
civil partnership     20.33%    442
unmarried             13.25%    288
divorced               5.15%    112
widow / widower        4.37%     95

### Preliminary Conclusion

This is the distribution of the dataset **(df_null)** categorized by the `'family_status'` column. Missing values in the `'days_employed'` and `'total_income'` columns are found in the following categories:

- married: 1237 rows
- civil partnership: 442 rows
- unmarried: 288 rows
- divorce: 112 rows
- widow / widower: 95 rows

In [13]:
# Examine other causes and patterns that may result in missing values in the filtered dataset
data_count = creditScore["income_type"].value_counts()
data_percent = creditScore["income_type"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : data_percent, "count" : data_count}).rename_axis("unique")

percent%  count
unique                                     
employee                      51.66%  11119
business                      23.62%   5085
retiree                       17.91%   3856
civil servant                  6.78%   1459
entrepreneur                   0.01%      2
unemployed                     0.01%      2
paternity / maternity leave     0.0%      1
student                         0.0%      1

### Preliminary Conclusion

This is the distribution of the dataset **(creditScore)** categorized based on the column `'income_type'`. The distribution of the data according to `'income_type'` is as follows:

- employee: 11119 rows
- business: 5085 rows
- retiree: 3856 rows
- civil servant: 1459 rows
- unemployed: 2 rows
- entrepreneur: 2 rows
- paternity / maternity leave: 1 row
- student: 1 row

In [14]:
# Check for other patterns - explain the patterns
# Create df_not_null

df_not_null = creditScore.dropna()

data_count = df_not_null["income_type"].value_counts()
data_percent = df_not_null["income_type"].value_counts(normalize=True).mul(100).round(2).astype(str)+"%"
pd.DataFrame(data={"percent%" : data_percent, "count" : data_count}).rename_axis("unique")

percent%  count
unique                                     
employee                      51.75%  10014
business                      23.65%   4577
retiree                       17.79%   3443
civil servant                  6.78%   1312
unemployed                     0.01%      2
entrepreneur                   0.01%      1
paternity / maternity leave    0.01%      1
student                        0.01%      1

`df_not_null` is the dataset without null values. This dataset can be used to understand the distribution of data without including rows containing null values.

The distribution of the dataset **(df_not_null)** categorized based on the column `'income_type'` is as follows:

- employee: 10014 rows
- business: 4577 rows
- retiree: 3443 rows
- civil servant: 1312 rows
- unemployed: 2 rows
- entrepreneur: 1 row
- paternity / maternity leave: 1 row
- student: 1 row

### Conclusion <a id='conclusion1'></a>

So the total missing data is 2174 rows. after we do data exploration from the three data distributions (**'creditScore', 'df_null', 'df_not_null'**) we can see that:

* the majority of customers are married people
* most of the customers are employees
* the most missing data is in the rows `'employee'` and `'married'` according to the `'income_type'` and `'family_status'` categories

This dataset provides enough information that we will need in our hypothesis testing. however, there are some missing data. if we look at the distribution based on categories, the missing data is in the majority group. we will fix this data in the next step, namely **Data Preprocessing**.

[back to content](#back)

# Stage 2. Data Preprocessing <a id='processing_data'></a>

In this stage, we will review, correct, and re-examine any anomalous data. It is advisable to correct these anomalous values first to avoid errors when testing the hypotheses we are going to evaluate.

### Data Transformation <a id='data_transform'>

In [15]:
# Let's review all values in the education column to check for any spelling that needs to be corrected.
creditScore['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

Oh no! Upon closer inspection, those values actually have the same meaning, but they are written in different styles. In terms of data processing, these anomalous values are referred to as `implicit duplicates`. If these values are included in our hypothesis testing, they will cause issues. We will address these anomalous values by converting them to lowercase.

In [16]:
# Correct the records if necessary.
creditScore['education'] = creditScore['education'].str.lower()

In [17]:
# Check all values in the column to ensure we have corrected them correctly
creditScore['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

Now, the values that were considered different earlier now have the same writing format. Thus, there are no more implicit duplicates in the column.

In [18]:
# Let's look at the distribution of values in the `children` column
creditScore['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

The logic dictates that humans cannot possibly have 20 children at once. It is also impossible for someone to have -1 child; this would mean they have no children. These errors may have occurred due to mistakes during data input. Therefore, we will correct these two values.

In [19]:
# Correct the data based on your decision
creditScore['children'].replace({-1: 1, 20: 2}, inplace=True)

In [20]:
# Recheck the `children` column to ensure that everything has been corrected

creditScore['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [21]:
# Find problematic data in the `days_employed` column if there are any issues and calculate the percentage
# Find the percentage of negative values for the number of days
(creditScore.loc[creditScore['days_employed'] < 0, 'days_employed'].count() / len(creditScore)) * 100

73.89547038327527

In [22]:
# Finding out the percentage of working days of customers exceeding their age.
(creditScore.loc[(creditScore['days_employed'] / 365) > creditScore['dob_years'], 'days_employed'].count() / len(creditScore)) * 100


16.004645760743323

It turns out that the number of negative values in the `'days_employed'` column is very high, around 73%. The number of unreasonable work experiences is also quite large, around 16%. Let's correct these two problematic values first.

In [23]:
# Resolve problematic values, if any
# Convert negative values to absolute values
creditScore['days_employed'] = round(abs(creditScore['days_employed']))

In [24]:
# Replace dob_years with 0
avg_age = int(creditScore['dob_years'].mean())
creditScore.loc[creditScore['dob_years']==0, 'dob_years'] = avg_age

In [25]:
# Convert days worked to years worked
creditScore['years_employed'] = round(creditScore['days_employed'] / 365)
creditScore.head()

children  days_employed  dob_years            education  education_id  \
0         1         8438.0         42    bachelor's degree             0   
1         1         4025.0         36  secondary education             1   
2         0         5623.0         33  secondary education             1   
3         3         4125.0         32  secondary education             1   
4         0       340266.0         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed  
0    purchase of the house            23.0  
1             car purchase            11.0  
2    purchase of the house            15.0  
3  supplementary education            11.0  
4        to have a wedding           932.0

In [26]:
# Get the actual duration of work
creditScore['year_working'] = creditScore['dob_years'] - 18
creditScore.head()

children  days_employed  dob_years            education  education_id  \
0         1         8438.0         42    bachelor's degree             0   
1         1         4025.0         36  secondary education             1   
2         0         5623.0         33  secondary education             1   
3         3         4125.0         32  secondary education             1   
4         0       340266.0         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  years_employed  year_working  
0    purchase of the house            23.0            24  
1             car purchase            11.0            18  
2    purchase of the house            15.0            15  
3  supplementary education            11.0            14  
4        to have a wedding           932.0            35

In [27]:
# Replace days_employed based on actual work experience
value = creditScore['year_working'] * 365
creditScore.loc[creditScore['years_employed'] > creditScore['year_working'], 'days_employed'] = value

In [28]:
# Check the results - make sure the issue has been fixed
creditScore.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.479721    4610.301638     43.495145      0.817236   
std        0.755528    5387.526183     12.218213      0.548138   
min        0.000000      24.000000     19.000000      0.000000   
25%        0.000000     924.000000     34.000000      1.000000   
50%        0.000000    2190.000000     43.000000      1.000000   
75%        1.000000    5503.000000     53.000000      1.000000   
max        5.000000   20440.000000     75.000000      4.000000   

       family_status_id          debt   total_income  years_employed  \
count      21525.000000  21525.000000   19351.000000    19351.000000   
mean           0.972544      0.080883   26787.568355      183.328820   
std            1.420324      0.272661   16475.450632      380.906543   
min            0.000000      0.000000    3306.762000        0.000000   
25%            0.000000      0.000000   16488.504500        3.000000   
50%            0.000000      0.000000   23202.870000        6.000000   
75%            1.000000      0.000000   32549.611000       15.000000   
max            4.000000      1.000000  362496.645000     1101.000000   

       year_working  
count  21525.000000  
mean      25.495145  
std       12.218213  
min        1.000000  
25%       16.000000  
50%       25.000000  
75%       35.000000  
max       57.000000

Great, now the `days_employed` values have been corrected. There are no more negative days and the work experience values are reasonable.

In [29]:
# Check `dob_years` for suspicious values and calculate their percentage
creditScore['dob_years'].sort_values().unique()

array([19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
       70, 71, 72, 73, 74, 75])

The problematic values are found in the age of customers being 0. This is likely due to errors during data input. We will replace these with the median or mean of the customers' ages.

In [30]:
creditScore['dob_years'].describe()

count    21525.000000
mean        43.495145
std         12.218213
min         19.000000
25%         34.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

It turns out that the mean and median values are not significantly different. We can use either one.

In [31]:
# Address issues in the `dob_years` column, if any
creditScore.loc[creditScore['dob_years'] == 0, 'dob_years'] = int(creditScore['dob_years'].mean())

In [32]:
# Check the results - make sure the issue has been fixed
creditScore['dob_years'].sort_values().unique()

array([19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
       70, 71, 72, 73, 74, 75])

Alright, now we will check the values in the `family_status` column and correct them if there are any errors.

In [33]:
# Let's examine the values for this column
creditScore['family_status'].unique()

array(['married', 'civil partnership', 'widow / widower', 'divorced',
       'unmarried'], dtype=object)

It seems that there are no problematic values in this column. We will skip this part.

Now, let's check the `gender` column. We will replace it if there are any problematic values.

In [34]:
# let's take a look at the values in this column.
creditScore['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

The entries in this column are actually not problematic. However, to clarify the meaning of each value, we can change them to values that are more descriptive. There is also one anomaly that we will remove using the `.drop()` method.

In [35]:
# Resolve problematic values, if any
creditScore.drop(creditScore[creditScore['gender'] == 'XNA'].index, inplace=True)
creditScore['gender'] = creditScore['gender'].replace({'F': 'Female', 'M': 'Male'})

In [36]:
# Let's fix any problematic values, if there are any.
creditScore['gender'].unique()

array(['Female', 'Male'], dtype=object)

Alright, we have corrected the values that were originally only 'F' and 'M' to 'Female' and 'Male', respectively. This clarifies the meaning of the data. We have also removed the 'XNA' value as it was an anomaly.

In [37]:
# Let's look at the values in this column
creditScore['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

It seems that there are no problematic values in this column. We will skip this part.

Now, after we have checked each column, we will look at the number of duplicates in our dataset.

### Fixing Duplicate Data <a id='duplicate'></a>

In [38]:
# Check for duplicates
creditScore.duplicated().sum()

71

There are 71 duplicates in the dataset. We will remove these duplicates to make the analysis more accurate. To do this, we will use the `.drop_duplicates()` method.

In [39]:
# Resolve duplicates, if any
creditScore = creditScore.drop_duplicates().reset_index(drop=True)

In [40]:
# Perform a final check to see if we have any duplicates
creditScore.duplicated().sum()

0

By removing duplicate rows, we have addressed the anomaly related to duplication. This helps ensure data consistency and accuracy when performing analysis.

In [41]:
# Check the size of the dataset after the first manipulation you performed
creditScore.shape

(21453, 14)

In [42]:
former_df = 21525
data_null_row = 2174
new_df = 21454

changes = (former_df - new_df) + data_null_row
percentage_change = (changes / former_df) * 100

print(f"Percentage change in data (including null): {percentage_change:.2f}%")

Percentage change in data (including null): 10.43%


The dataset we have has undergone a process of correcting the writing and has been manipulated in the 'days_employeed' column. The percentage change is around 10.43%.

### Fixing Missing Values <a id='missing_values'></a>

Sometimes, to make it easier for us to fix data, we can use a dictionary. In this dataset, `'family_status_id'` and `'family_status'` are actually the same values. The same goes for `'education_id'` and `'education'`. We will change those columns into dictionaries.

In [43]:
# Find the dictionary
family_id = dict(zip(creditScore.family_status_id, creditScore.family_status))
education_id = dict(zip( creditScore.education_id, creditScore.education ))

In [44]:
family_id

{0: 'married',
 1: 'civil partnership',
 2: 'widow / widower',
 3: 'divorced',
 4: 'unmarried'}

In [45]:
education_id

{0: "bachelor's degree",
 1: 'secondary education',
 2: 'some college',
 3: 'primary education',
 4: 'graduate degree'}

### Categorizing Age and Occupation as Parameters

The main factors in identifying total income are usually age and occupation. Let's use these two factors as parameters to fill in the missing values in the 'total_income' column.

In [46]:
# Let's write a function to calculate age categories
def age_category(age):
    if age <= 25:
        return '0-25 y/o'
    elif age > 25 and age <= 50:
        return '26-50 y/o'
    else:
        return '50+ y/o'

In [47]:
# Let's perform a test to see if your function works or not.
print('client age is:', age_category(35))

client age is: 26-50 y/o


In [48]:
# Let's create a new column based on the function
creditScore['age_category'] = creditScore['dob_years'].apply(age_category)

In [49]:
# Check the values in the new column
creditScore['age_category']

0        26-50 y/o
1        26-50 y/o
2        26-50 y/o
3        26-50 y/o
4          50+ y/o
           ...    
21448    26-50 y/o
21449      50+ y/o
21450    26-50 y/o
21451    26-50 y/o
21452    26-50 y/o
Name: age_category, Length: 21453, dtype: object

### Fixing Values in the 'total_income' Column

Earlier, we saw that the values in the 'total_income' column are missing. Let's fix those values first.

In [50]:
# Create a table without missing values and display some rows to ensure everything is running smoothly.
not_null_income = creditScore.loc[-creditScore['total_income'].isnull()]

In [51]:
# Note the average income values based on the factors you have identified
mean_income = pd.pivot_table(
    not_null_income,
    index = ['income_type', 'age_category'],
    values = 'total_income',
    aggfunc = 'mean'
).reset_index()

mean_income.head(10)

income_type age_category  total_income
0       business     0-25 y/o  25713.363026
1       business    26-50 y/o  33044.461867
2       business      50+ y/o  32442.468551
3  civil servant     0-25 y/o  24510.974242
4  civil servant    26-50 y/o  27588.263377
5  civil servant      50+ y/o  27113.556702
6       employee     0-25 y/o  22248.409440
7       employee    26-50 y/o  26015.220249
8       employee      50+ y/o  26408.949768
9   entrepreneur    26-50 y/o  79866.103000

In [52]:
# Note the median income values based on the factors you have identified
median_income = pd.pivot_table(
    not_null_income,
    index = ['income_type', 'age_category'],
    values = 'total_income',
    aggfunc = 'median'
).reset_index()

median_income.head(10)

income_type age_category  total_income
0       business     0-25 y/o    22814.5995
1       business    26-50 y/o    28372.6700
2       business      50+ y/o    27486.8960
3  civil servant     0-25 y/o    22758.5535
4  civil servant    26-50 y/o    24328.8550
5  civil servant      50+ y/o    23585.3185
6       employee     0-25 y/o    20599.1375
7       employee    26-50 y/o    23020.9075
8       employee      50+ y/o    22869.8610
9   entrepreneur    26-50 y/o    79866.1030

If we think about it again, occupation greatly influences total income. However, the possibility of outlier values is very high. It would be more appropriate to use the median value as a reference.

In [53]:
#  Write the function we will use to fill in missing values
def replace_income(dataset, grouping):
    for _, row in grouping.iterrows():
        income_type = row['income_type']
        age_category = row['age_category']
        median_income_value = row['total_income']

        dataset.loc[(dataset['income_type'] == income_type) & (dataset['age_category'] == age_category) & (dataset['total_income'].isnull()), 'total_income'] = median_income_value

    return dataset

In [54]:
# Check the values in the new column
creditScore = replace_income(dataset=creditScore, grouping=median_income)
creditScore

children  days_employed  dob_years            education  education_id  \
0             1         8438.0         42    bachelor's degree             0   
1             1         4025.0         36  secondary education             1   
2             0         5623.0         33  secondary education             1   
3             3         4125.0         32  secondary education             1   
4             0        12775.0         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1         4529.0         43  secondary education             1   
21449         0        17885.0         67  secondary education             1   
21450         1         2113.0         38  secondary education             1   
21451         3         3112.0         38  secondary education             1   
21452         2         1985.0         40  secondary education             1   

           family_status  family_status_id  gender income_type  debt  \
0                married                 0  Female    employee     0   
1                married                 0  Female    employee     0   
2                married                 0    Male    employee     0   
3                married                 0    Male    employee     0   
4      civil partnership                 1  Female     retiree     0   
...                  ...               ...     ...         ...   ...   
21448  civil partnership                 1  Female    business     0   
21449            married                 0  Female     retiree     0   
21450  civil partnership                 1    Male    employee     1   
21451            married                 0    Male    employee     1   
21452            married                 0  Female    employee     0   

       total_income                  purpose  years_employed  year_working  \
0         40620.102    purchase of the house            23.0            24   
1         17932.802             car purchase            11.0            18   
2         23341.752    purchase of the house            15.0            15   
3         42820.568  supplementary education            11.0            14   
4         25378.572        to have a wedding           932.0            35   
...             ...                      ...             ...           ...   
21448     35966.698     housing transactions            12.0            25   
21449     24959.969        purchase of a car           942.0            49   
21450     14347.610                 property             6.0            20   
21451     39054.888        buying my own car             9.0            20   
21452     13127.587             to buy a car             5.0            22   

      age_category  
0        26-50 y/o  
1        26-50 y/o  
2        26-50 y/o  
3        26-50 y/o  
4          50+ y/o  
...            ...  
21448    26-50 y/o  
21449      50+ y/o  
21450    26-50 y/o  
21451    26-50 y/o  
21452    26-50 y/o  

[21453 rows x 15 columns]

It looks like we have fixed the missing values.

In [55]:
# Check if we get any errors
creditScore.isnull().sum()

children               0
days_employed       2103
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           1
purpose                0
years_employed      2103
year_working           0
age_category           0
dtype: int64

It seems there is one value that cannot be filled with the conditions in the function. Let's see what the cause is.

In [56]:
creditScore[creditScore['total_income'].isnull()]

children  days_employed  dob_years          education  education_id  \
5931         0            NaN         58  bachelor's degree             0   

     family_status  family_status_id gender   income_type  debt  total_income  \
5931       married                 0   Male  entrepreneur     0           NaN   

                          purpose  years_employed  year_working age_category  
5931  buy residential real estate             NaN            40      50+ y/o

In [57]:
creditScore[creditScore['income_type'] == 'entrepreneur']

children  days_employed  dob_years          education  education_id  \
5931          0            NaN         58  bachelor's degree             0   
18646         0          521.0         27  bachelor's degree             0   

           family_status  family_status_id  gender   income_type  debt  \
5931             married                 0    Male  entrepreneur     0   
18646  civil partnership                 1  Female  entrepreneur     0   

       total_income                      purpose  years_employed  \
5931            NaN  buy residential real estate             NaN   
18646     79866.103             having a wedding             1.0   

       year_working age_category  
5931             40      50+ y/o  
18646             9    26-50 y/o

After investigating, it turns out that the condition on row 5931 does not meet the conditions in the function. This is because the 'median_income' dataset does not calculate the total income for customers who are over 50 years old. So the condition `income_type` == `'entrepreneur'` and `age_category` == `'50+ y/o'` does not apply. Since this value is an anomaly, we will just delete it.

In [58]:
creditScore.dropna(subset=['total_income'], inplace=True)

In [59]:
# recheck problematic values
creditScore.isnull().sum()

children               0
days_employed       2102
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
years_employed      2102
year_working           0
age_category           0
dtype: int64

### Fixing values in `days_employed`

The values in the `days_employed` column indicate how long customers have been working in days. When filling in the missing values in this column, it makes sense to use the age of the customer as a parameter.

In [60]:
# Check the number of entries in the column
print(creditScore['days_employed'].size)
print(creditScore.groupby('age_category')['days_employed'].count())

21452
age_category
0-25 y/o      1112
26-50 y/o    12268
50+ y/o       5970
Name: days_employed, dtype: int64


In the entries above, there are `21453` rows of customer work experience and 3 age categories. With this, we can fill in the missing values in the `'days_employed'` column.

In [61]:
# Median distribution of `days_employed` based on the parameters you identified
median_days = pd.pivot_table(
    creditScore,
    index = 'age_category',
    values = 'days_employed',
    aggfunc = 'median',
).reset_index()

median_days.head()

age_category  days_employed
0     0-25 y/o          771.5
1    26-50 y/o         1700.0
2      50+ y/o        12775.0

In [62]:
# Mean distribution of `days_employed` based on the parameters you identified
mean_days = pd.pivot_table(
    creditScore,
    index = 'age_category',
    values = 'days_employed',
    aggfunc = 'mean',
).reset_index()

mean_days.head()

age_category  days_employed
0     0-25 y/o     889.911871
1    26-50 y/o    2385.710548
2      50+ y/o    9875.060134

If we assume that people start working at the age of 19, then it would make sense to use the mean value.

In [63]:
# Let's write a function that calculates the mean or median (depending on your decision) based on the parameters you identified
def fill_days(dataset, parameter):
    mean_values = dataset.groupby(parameter)['days_employed'].mean()
    dataset['days_employed'] = dataset['days_employed'].fillna(dataset[parameter].apply(lambda x: mean_values[x]))
    return dataset

In [64]:
# Check if your function works
fill_days(creditScore, 'age_category')

children  days_employed  dob_years            education  education_id  \
0             1         8438.0         42    bachelor's degree             0   
1             1         4025.0         36  secondary education             1   
2             0         5623.0         33  secondary education             1   
3             3         4125.0         32  secondary education             1   
4             0        12775.0         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1         4529.0         43  secondary education             1   
21449         0        17885.0         67  secondary education             1   
21450         1         2113.0         38  secondary education             1   
21451         3         3112.0         38  secondary education             1   
21452         2         1985.0         40  secondary education             1   

           family_status  family_status_id  gender income_type  debt  \
0                married                 0  Female    employee     0   
1                married                 0  Female    employee     0   
2                married                 0    Male    employee     0   
3                married                 0    Male    employee     0   
4      civil partnership                 1  Female     retiree     0   
...                  ...               ...     ...         ...   ...   
21448  civil partnership                 1  Female    business     0   
21449            married                 0  Female     retiree     0   
21450  civil partnership                 1    Male    employee     1   
21451            married                 0    Male    employee     1   
21452            married                 0  Female    employee     0   

       total_income                  purpose  years_employed  year_working  \
0         40620.102    purchase of the house            23.0            24   
1         17932.802             car purchase            11.0            18   
2         23341.752    purchase of the house            15.0            15   
3         42820.568  supplementary education            11.0            14   
4         25378.572        to have a wedding           932.0            35   
...             ...                      ...             ...           ...   
21448     35966.698     housing transactions            12.0            25   
21449     24959.969        purchase of a car           942.0            49   
21450     14347.610                 property             6.0            20   
21451     39054.888        buying my own car             9.0            20   
21452     13127.587             to buy a car             5.0            22   

      age_category  
0        26-50 y/o  
1        26-50 y/o  
2        26-50 y/o  
3        26-50 y/o  
4          50+ y/o  
...            ...  
21448    26-50 y/o  
21449      50+ y/o  
21450    26-50 y/o  
21451    26-50 y/o  
21452    26-50 y/o  

[21452 rows x 15 columns]

In [65]:
# Apply the function to income_type
fill_days(creditScore, 'income_type')

children  days_employed  dob_years            education  education_id  \
0             1         8438.0         42    bachelor's degree             0   
1             1         4025.0         36  secondary education             1   
2             0         5623.0         33  secondary education             1   
3             3         4125.0         32  secondary education             1   
4             0        12775.0         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1         4529.0         43  secondary education             1   
21449         0        17885.0         67  secondary education             1   
21450         1         2113.0         38  secondary education             1   
21451         3         3112.0         38  secondary education             1   
21452         2         1985.0         40  secondary education             1   

           family_status  family_status_id  gender income_type  debt  \
0                married                 0  Female    employee     0   
1                married                 0  Female    employee     0   
2                married                 0    Male    employee     0   
3                married                 0    Male    employee     0   
4      civil partnership                 1  Female     retiree     0   
...                  ...               ...     ...         ...   ...   
21448  civil partnership                 1  Female    business     0   
21449            married                 0  Female     retiree     0   
21450  civil partnership                 1    Male    employee     1   
21451            married                 0    Male    employee     1   
21452            married                 0  Female    employee     0   

       total_income                  purpose  years_employed  year_working  \
0         40620.102    purchase of the house            23.0            24   
1         17932.802             car purchase            11.0            18   
2         23341.752    purchase of the house            15.0            15   
3         42820.568  supplementary education            11.0            14   
4         25378.572        to have a wedding           932.0            35   
...             ...                      ...             ...           ...   
21448     35966.698     housing transactions            12.0            25   
21449     24959.969        purchase of a car           942.0            49   
21450     14347.610                 property             6.0            20   
21451     39054.888        buying my own car             9.0            20   
21452     13127.587             to buy a car             5.0            22   

      age_category  
0        26-50 y/o  
1        26-50 y/o  
2        26-50 y/o  
3        26-50 y/o  
4          50+ y/o  
...            ...  
21448    26-50 y/o  
21449      50+ y/o  
21450    26-50 y/o  
21451    26-50 y/o  
21452    26-50 y/o  

[21452 rows x 15 columns]

In [66]:
# Check if your function works
creditScore['days_employed'].isna().sum()

0

In [67]:
# Check the entries in all columns - make sure we fix all missing values
creditScore.isna().sum()

children               0
days_employed          0
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
years_employed      2102
year_working           0
age_category           0
dtype: int64

In [68]:
# fill in the missing 'years_employed' column
creditScore['years_employed'] = creditScore['days_employed'] / 365

In [69]:
creditScore.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
years_employed      0
year_working        0
age_category        0
dtype: int64

### Data Categorization <a id='data_category'></a>

In some columns with a large number of unique values, we can create data categorization for those columns. This aims to facilitate us in reviewing the unique values in the dataset.

In [70]:
# create a new DataFrame
data = creditScore.drop(['year_working', 'years_employed'], axis=1)

In [71]:
# view data categories from the `purpose` column
data['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

If we look at it, the unique values in the `purpose` column vary greatly. However, some actually have the same meaning. So we will categorize its values, but make it easier to read.

In [72]:
# create a function that categorizes the 'purpose' column

def purpose_category(purpose):
    category= {
        'housing': ['house', 'housing'],
        'car': ['car'],
        'education': ['education', 'university', 'educated'],
        'wedding': ['wedding'],
        'real_estate': ['real_estate']
    }

    for key, values in category.items():
        for value in values:
            if value in purpose:
                return key
    return 'other'

In [73]:
# Applying a function to the purpose column.
data['purpose_category'] = data['purpose'].apply(purpose_category)
data

children  days_employed  dob_years            education  education_id  \
0             1         8438.0         42    bachelor's degree             0   
1             1         4025.0         36  secondary education             1   
2             0         5623.0         33  secondary education             1   
3             3         4125.0         32  secondary education             1   
4             0        12775.0         53  secondary education             1   
...         ...            ...        ...                  ...           ...   
21448         1         4529.0         43  secondary education             1   
21449         0        17885.0         67  secondary education             1   
21450         1         2113.0         38  secondary education             1   
21451         3         3112.0         38  secondary education             1   
21452         2         1985.0         40  secondary education             1   

           family_status  family_status_id  gender income_type  debt  \
0                married                 0  Female    employee     0   
1                married                 0  Female    employee     0   
2                married                 0    Male    employee     0   
3                married                 0    Male    employee     0   
4      civil partnership                 1  Female     retiree     0   
...                  ...               ...     ...         ...   ...   
21448  civil partnership                 1  Female    business     0   
21449            married                 0  Female     retiree     0   
21450  civil partnership                 1    Male    employee     1   
21451            married                 0    Male    employee     1   
21452            married                 0  Female    employee     0   

       total_income                  purpose age_category purpose_category  
0         40620.102    purchase of the house    26-50 y/o          housing  
1         17932.802             car purchase    26-50 y/o              car  
2         23341.752    purchase of the house    26-50 y/o          housing  
3         42820.568  supplementary education    26-50 y/o        education  
4         25378.572        to have a wedding      50+ y/o          wedding  
...             ...                      ...          ...              ...  
21448     35966.698     housing transactions    26-50 y/o          housing  
21449     24959.969        purchase of a car      50+ y/o              car  
21450     14347.610                 property    26-50 y/o            other  
21451     39054.888        buying my own car    26-50 y/o              car  
21452     13127.587             to buy a car    26-50 y/o              car  

[21452 rows x 14 columns]

In [74]:
# Checking the category of the 'purpose' column.
data['purpose_category'].unique()

array(['housing', 'car', 'education', 'wedding', 'other'], dtype=object)

The category of the 'purpose' column has been updated. Now it's time to categorize the numerical data.

In [75]:
# View all numerical data in the column you selected for categorization.
data['total_income'].describe()

count     21452.000000
mean      26454.931513
std       15709.205150
min        3306.762000
25%       17211.355500
50%       23020.907500
75%       31328.693750
max      362496.645000
Name: total_income, dtype: float64

Got it! Based on the description you provided, we'll create a function to categorize the data based on total income ranges. This categorization will help us examine the hypothesis of defaulting. The smallest value in the data is 3306, and the largest value is 362496. Let's proceed with creating the function to categorize the data into income ranges.

In [76]:
# Create a function that categorizes data into different numerical groups based on ranges
def income_category(income):
    if income <= 15000:
        value = '<= 15k'
    elif income > 15000 and income <= 30000 :
        value = '15-30k'
    elif income > 30000 and income <= 45000 :
        value = '30-45k'
    elif income > 45000 and income <= 60000 :
        value = '45-60k'
    else:
        value = '> 60k'
    return value

In [77]:
# Create a column containing categories.
data['income_category'] = data['total_income'].apply(income_category)

In [78]:
# Calculate each category value to see its distribution
data['income_category'].value_counts()

15-30k    11790
30-45k     4059
<= 15k     3744
45-60k     1187
> 60k       672
Name: income_category, dtype: int64

# Step 3. Testing Hypotheses <a id='hypothesis'></a>
Finally, we've reached the last step! We'll examine whether there is a relationship between the following four points:

- the number of children and the likelihood of defaulting on a loan
- testing whether there is a relationship between marital status and the probability of defaulting on a loan
- whether an individual's income level can affect loan default
- determining if the purpose of a loan affects the likelihood of defaulting on it.

**Is there a correlation between having children and the probability of defaulting on a loan?**

In [79]:
# Check the data for children and loan default.
children_debt = pd.pivot_table(data, index='children', columns='debt', values='education', aggfunc='count').fillna(0)
children_debt.sort_values(1, ascending=False)

debt            0       1
children                 
0         13026.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     0.0

In [80]:
# Calculate the default percentage based on the number of children.
children_debt['total'] = children_debt[0] + children_debt[1]
children_debt['percent 0'] = round(children_debt[0] / children_debt['total'] * 100, 1).astype(str) + '%'
children_debt['percent 1'] = round(children_debt[1] / children_debt['total'] * 100, 1).astype(str) + '%'
children_debt.sort_values(1, ascending=False)

debt            0       1    total percent 0 percent 1
children                                              
0         13026.0  1063.0  14089.0     92.5%      7.5%
1          4410.0   445.0   4855.0     90.8%      9.2%
2          1926.0   202.0   2128.0     90.5%      9.5%
3           303.0    27.0    330.0     91.8%      8.2%
4            37.0     4.0     41.0     90.2%      9.8%
5             9.0     0.0      9.0    100.0%      0.0%

**Conclusion**

Logically, if we have more children, the percentage of customer defaults will increase. Based on the data above, when looking at the number of defaulters according to their category, the default percentage shows a decreasing correlation, meaning that the more children a customer has, the fewer defaulters there are.

However, there's something we need to pay attention to here. Most groups indeed consist of a majority of customers with fewer children. This means that the conclusion above is not entirely accurate. Let's look at the default percentage for each group. The group of customers with 4 children has a high default percentage.

**Is there a correlation between marital status and the probability of defaulting on a loan?**

In [81]:
# Check the data for marital status and loan default.
family_debt = pd.pivot_table(data, index= 'family_status', columns= 'debt', values= 'children',aggfunc= 'count').fillna(0)
family_debt.sort_values(1, ascending=False)

debt                   0    1
family_status                
married            11407  931
civil partnership   3762  388
unmarried           2536  274
divorced            1110   85
widow / widower      896   63

In [82]:
# Calculate the default percentage based on marital status.
family_debt['total'] = family_debt[0] + family_debt[1]
family_debt['percent 0'] = round(family_debt[0] / family_debt['total'] * 100, 1).astype(str) + '%'
family_debt['percent 1'] = round(family_debt[1] / family_debt['total'] * 100, 1).astype(str) + '%'
family_debt.sort_values(1, ascending=False)

debt                   0    1  total percent 0 percent 1
family_status                                           
married            11407  931  12338     92.5%      7.5%
civil partnership   3762  388   4150     90.7%      9.3%
unmarried           2536  274   2810     90.2%      9.8%
divorced            1110   85   1195     92.9%      7.1%
widow / widower      896   63    959     93.4%      6.6%

**Conclusion**

From the calculations, it can be concluded that the marital status "widow/widower" has the lowest percentage of loan defaults (6.6%), followed by "divorced" (7.1%) and "married" (7.5%). Meanwhile, the marital statuses "civil partnership" and "unmarried" have higher percentages of loan defaults, which are 9.3% and 9.8% respectively.

**Is there a correlation between income level and the probability of defaulting on a loan?**

In [83]:
# Calculate the default percentage based on income level.
income_type_debt = pd.pivot_table(data, index= 'income_category', columns= 'debt',values= 'children',aggfunc= 'count').fillna(0)
income_type_debt.sort_values(1, ascending=False)

debt                 0     1
income_category             
15-30k           10783  1007
30-45k            3755   304
<= 15k            3446   298
45-60k            1093    94
> 60k              634    38

In [84]:
# Calculate the default percentage based on income level.
income_type_debt['total'] = income_type_debt[0] + income_type_debt[1]
income_type_debt['percent 0'] = round(income_type_debt[0] / income_type_debt['total'] * 100, 1).astype(str) + '%'
income_type_debt['percent 1'] = round(income_type_debt[1] / income_type_debt['total'] * 100, 1).astype(str) + '%'
income_type_debt.sort_values(1, ascending=False)

debt                 0     1  total percent 0 percent 1
income_category                                        
15-30k           10783  1007  11790     91.5%      8.5%
30-45k            3755   304   4059     92.5%      7.5%
<= 15k            3446   298   3744     92.0%      8.0%
45-60k            1093    94   1187     92.1%      7.9%
> 60k              634    38    672     94.3%      5.7%

**Conclusion**

It can be concluded that the "> 60k" income group has the lowest percentage of loan defaults (5.7%), followed by the "30-45k" (7.5%) and "45-60k" (7.9%) income groups. Meanwhile, the "15-30k" and "<= 15k" income groups have higher percentages of loan defaults, which are 8.5% and 8.0% respectively.

**How does the purpose of the loan affect the default percentage?**

In [85]:
purpose_debt = pd.pivot_table(data, index= 'purpose_category', columns= 'debt', values= 'total_income',aggfunc= 'count').fillna(0)

purpose_debt = pd.pivot_table(data, index= 'purpose_category', columns= 'debt',values= 'children',aggfunc= 'count').fillna(0)
purpose_debt.sort_values(1, ascending=False)

debt                 0    1
purpose_category           
other             6474  526
car               3903  403
education         3643  370
housing           3553  256
wedding           2138  186

In [86]:
# Check the default percentage for each loan purpose and perform analysis.
purpose_debt = pd.pivot_table(data, index= 'purpose_category', columns= 'debt', values= 'total_income',aggfunc= 'count').fillna(0)

purpose_debt['total'] = purpose_debt[0] + purpose_debt[1]
purpose_debt['percent 0'] = round(purpose_debt[0] / purpose_debt['total'] * 100, 1).astype(str) + '%'
purpose_debt['percent 1'] = round(purpose_debt[1] / purpose_debt['total'] * 100, 1).astype(str) + '%'
purpose_debt.sort_values(1, ascending=False)

debt                 0    1  total percent 0 percent 1
purpose_category                                      
other             6474  526   7000     92.5%      7.5%
car               3903  403   4306     90.6%      9.4%
education         3643  370   4013     90.8%      9.2%
housing           3553  256   3809     93.3%      6.7%
wedding           2138  186   2324     92.0%      8.0%

#### **Conclusion**

The loan purpose category "housing" has the lowest percentage of loan defaults (6.7%), followed by the "other" (7.5%) and "wedding" (8.0%) categories. Meanwhile, the "car" and "education" loan purpose categories have higher percentages of loan defaults, which are 9.4% and 9.2% respectively.

# General Conclusion <a id='final'></a>

**We've completed three stages of data analysis:**
* Data Review
* Data Preprocessing
* Testing Hypotheses

In the Data Review stage, we explored the distribution of customer data. We found some missing data which we then addressed in the Data Preprocessing stage. In this stage, besides filling in the missing data, we also corrected implicit duplicates in the 'purpose' and 'education' columns. The updated dataset was then used in our Testing Hypotheses stage, where we tested 4 hypotheses:

1. the relationship between the number of children and loan default.
2. whether there's a relationship between marital status and the probability of defaulting on a loan.
3. whether an individual's income level affects loan default.
4. whether the purpose of a loan affects the likelihood of defaulting on it.

**After testing the hypotheses, we can conclude that:**

1. Having more children doesn't necessarily lead to loan default.
2. The highest increase in default occurs for those who are **never married**, namely 'unmarried' and 'civil partnership'.
3. This hypothesis is accepted because there's an increase in default percentage for lower and lower-middle income.
4. The loan purposes with the highest default rates are for car and education purposes.